### PR TITLE
Pynumero interior point

### PR DESCRIPTION
**Pull request by michaelbynum**
@Robbybp There is still work to do, but I'd like your input on the changes I have been making. I tried to simplify a lot of code (both in the `InteriorPointSolver` and the `InteriorPointInterface`). 
1. I added the `LinearSolverResults` object as we discussed. 
2. I removed all of the compression matrix stuff to improve performance. 
3. I was having trouble with the the second half of the regularization test. I was getting errors because the regularization coefficient was getting to large, so I cut the second half of that test (the first half was perfect), and added a second test.
4. I modified the logging slightly to log every iteration of the factorization rather than just iterations with regularization (still with debug).

### CODE REVIEW

**Comment by Robbybp** [comment]
Changes look great at a quick scan. I'll try to look through this in detail and play around with it tomorrow evening.

**Comment by Robbybp** [comment]
> I was having trouble with the the second half of the regularization test. I was getting errors because the regularization coefficient was getting to large, so I cut the second half of that test (the first half was perfect), and added a second test.
 
Before checking out your branch to test/inspect, I pulled from pyomo:master and ran these tests, and got the same error. Not sure what could have changed that is making the algorithm unable to construct or Mumps unable to recognize a matrix with correct inertia.

**Comment by Robbybp** [comment]
> Before checking out your branch to test/inspect, I pulled from pyomo:master and ran these tests, and got the same error. 

My scipy regularization test seems to be working fine, FWIW.

**Comment by michaelbynum** [comment]
This test problem is really ugly. Without a line search, the algorithm takes steps that are crazy. In the second or third iteration, I see infeasibilities around 1e50. Given the enormous values, there are probably numerical issues that make it very difficult to predict what is going to happen.

File: pyomo/contrib/interior_point/linalg/tests/test_linear_solvers.py [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+class TestWrongNonzeroOrdering(unittest.TestCase):
+    def _test_solvers(self, solver, use_tril):
+        mat = get_base_matrix(use_tril=use_tril)
+        wrong_order_mat = get_base_matrix_wrong_order(use_tril=use_tril)
+        stat = solver.do_symbolic_factorization(mat)
+        stat = solver.do_numeric_factorization(wrong_order_mat)
+        x_true = np.array([1, 2, 3], dtype=np.double)
+        rhs = mat * x_true
+        x = solver.do_back_solve(rhs)
+        self.assertTrue(np.allclose(x, x_true))
```
Comment (Robbybp) [inline pyomo/contrib/interior_point/linalg/tests/test_linear_solvers.py:106]: What is the purpose of this test?

**Comment by Robbybp** [comment]
> This test problem is really ugly. Without a line search, the algorithm takes steps that are crazy. In the second or third iteration, I see infeasibilities around 1e50. Given the enormous values, there are probably numerical issues that make it very difficult to predict what is going to happen.

Yep that matches my observations. Not surprising, I guess, since this is basically the textbook example of degenerate equality constraints.

**Comment by Robbybp** [comment]
Would you be opposed if I changed the FactorizationContext to log to a `linear_solver_log_file` as long as one is provided? 

Just read and tested the code and it looks good to me. What do you plan to work on next, out of curiosity?

**Pull request closed without merging**